### ISSUE TITLE
 pvget et al. crash during startup

### ISSUE THREAD

**Issue by mdavidsaver** [epics-base/pvAccessCPP#95]
As a sort of complement to #92, a crash on startup:

```
Program terminated with signal 11, Segmentation fault.
#0  (anonymous namespace)::SearchResponseHandler::handleResponse (this=0x1,  
Missing separate debuginfos, use: debuginfo-install libgcc-4.4.7-18.el6_9.2.x86_64 l
ibstdc++-4.4.7-18.el6_9.2.x86_64 ncurses-libs-5.7-4.20090207.el6.x86_64 readline-6.0
-4.el6.x86_64
   responseFrom=<value optimized out>, transport=..., version=1 '\001',  
   command=<value optimized out>, payloadSize=<value optimized out>,  
   payloadBuffer=0xbe7580) at ../../src/remoteClient/clientContextImpl.cpp:2595
2595                csm->searchResponse(guid, cid, searchSequenceId, version, &serve
rAddress);
(gdb) thread apply all backtrace

Thread 3 (Thread 0x7f07631fd700 (LWP 14564)):
#0  pthread_cond_wait@@GLIBC_2.3.2 ()
   at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:183
#1  0x0000000000631756 in epicsEventWait (pevent=0xbe65a0)
   at ../../../src/libCom/osi/os/posix/osdEvent.c:103
#2  0x0000000000590150 in epics::pvData::Event::wait (this=<value optimized out>)
   at ../../src/misc/event.cpp:53
#3  0x00000000005912e2 in epics::pvData::Timer::run (this=0xbe64b0)
   at ../../src/misc/timer.cpp:135
#4  0x000000000062a4c9 in epicsThreadCallEntryPoint (pPvt=0xbe64e0)
   at ../../../src/libCom/osi/epicsThread.cpp:83
#5  0x000000000062fa6c in start_routine (arg=0xbe6760)
   at ../../../src/libCom/osi/os/posix/osdThread.c:403
#6  0x00000033af807aa1 in start_thread (arg=0x7f07631fd700) at pthread_create.c:301
#7  0x00000033af0e8bcd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:115

Thread 2 (Thread 0x7f0763c8f720 (LWP 14558)):
#0  0x000000000055856a in setIgnoredAddresses (serverFlag=<value optimized out>,  
   udpTransports=std::vector of length 1, capacity 1 = {...},  
   ifaceList=std::vector of length 2, capacity 2 = {...}, responseHandler=std::tr1::shared_ptr (count 4) 0xbe63d0, sendTransport=<value optimized out>,  
   listenPort=@0xbe58d4, autoAddressList=false, addressList=
   "aird-pc90626.slac.stanford.edu", ignoreAddressList="")
   at ../../src/remote/pv/blockingUDP.h:234
#1  epics::pvAccess::initializeUDPTransports (serverFlag=<value optimized out>,  
   udpTransports=std::vector of length 1, capacity 1 = {...},  
   ifaceList=std::vector of length 2, capacity 2 = {...}, responseHandler=std::tr1::shared_ptr (count 4) 0xbe63d0, sendTransport=<value optimized out>,  
   listenPort=@0xbe58d4, autoAddressList=false, addressList=
   "aird-pc90626.slac.stanford.edu", ignoreAddressList="")
   at ../../src/remote/blockingUDPTransport.cpp:746
#2  0x0000000000535575 in initializeUDPTransport (this=0xbe5890)
   at ../../src/remoteClient/clientContextImpl.cpp:4188
#3  internalInitialize (this=0xbe5890)
   at ../../src/remoteClient/clientContextImpl.cpp:4158
#4  (anonymous namespace)::InternalClientContextImpl::initialize (this=0xbe5890)
   at ../../src/remoteClient/clientContextImpl.cpp:4064
#5  0x0000000000531f7a in epics::pvAccess::createClientProvider (
   conf=<value optimized out>)
   at ../../src/remoteClient/clientContextImpl.cpp:4829
#6  0x00000000004fdc8c in epics::pvAccess::(anonymous namespace)::FunctionFactory::s
haredInstance (this=0xbe4cb0) at ../../src/factory/ChannelAccessFactory.cpp:104
#7  0x00000000004fd1d4 in epics::pvAccess::ChannelProviderRegistry::getProvider (
   this=<value optimized out>, providerName=<value optimized out>)
   at ../../src/factory/ChannelAccessFactory.cpp:42
#8  0x00000000004c9c7c in main (argc=<value optimized out>,  
   argv=<value optimized out>) at ../pvinfo.cpp:171

Thread 1 (Thread 0x7f0762f49700 (LWP 14574)):
#0  (anonymous namespace)::SearchResponseHandler::handleResponse (this=0x1,  
   responseFrom=<value optimized out>, transport=
   std::tr1::shared_ptr (count 2) 0xbe7470, version=1 '\001',  
   command=<value optimized out>, payloadSize=<value optimized out>,  
   payloadBuffer=0xbe7580) at ../../src/remoteClient/clientContextImpl.cpp:2595
#1  0x00000000005346cd in (anonymous namespace)::ClientResponseHandler::handleRespon
se (this=<value optimized out>, responseFrom=<value optimized out>,  
   transport=<value optimized out>, version=<value optimized out>,  
   command=<value optimized out>, payloadSize=<value optimized out>,  
   payloadBuffer=0xbe7580) at ../../src/remoteClient/clientContextImpl.cpp:2997
---Type <return> to continue, or q <return> to quit---
#2  0x000000000055557a in epics::pvAccess::BlockingUDPTransport::processBuffer (
   this=0xbe7470, transport=std::tr1::shared_ptr (count 2) 0xbe7470,  
   fromAddress=..., receiveBuffer=0xbe7580)
   at ../../src/remote/blockingUDPTransport.cpp:383
#3  0x000000000055771c in epics::pvAccess::BlockingUDPTransport::run (
   this=0xbe7470) at ../../src/remote/blockingUDPTransport.cpp:240
#4  0x000000000062a4c9 in epicsThreadCallEntryPoint (pPvt=0xc09390)
   at ../../../src/libCom/osi/epicsThread.cpp:83
#5  0x000000000062fa6c in start_routine (arg=0xc094f0)
   at ../../../src/libCom/osi/os/posix/osdThread.c:403
#6  0x00000033af807aa1 in start_thread (arg=0x7f0762f49700) at pthread_create.c:301
#7  0x00000033af0e8bcd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:115
```

**Comment by mdavidsaver** [comment]
This is one of the many places which valgrind flags as a possible race

```
$ valgrind --tool=helgrind ./bin/linux-x86_64-debug/pvget cnt
...
Possible data race during read of size 8 at 0x759E8C0 by thread #4
Locks held: none
   at 0x51F9D0A: (anonymous namespace)::SearchResponseHandler::handleResponse(osiSockAddr*, std::shared_ptr<epics::pvAccess::Transport> const&, signed char, signed char, unsigned long, epics::pvData::ByteBuffer*) (clientContextImpl.cpp:2595)
   by 0x51FC43C: (anonymous namespace)::ClientResponseHandler::handleResponse(osiSockAddr*, std::shared_ptr<epics::pvAccess::Transport> const&, signed char, signed char, unsigned long, epics::pvData::ByteBuffer*) (clientContextImpl.cpp:2997)
   by 0x51BBAE6: epics::pvAccess::BlockingUDPTransport::processBuffer(std::shared_ptr<epics::pvAccess::Transport> const&, osiSockAddr&, epics::pvData::ByteBuffer*) (blockingUDPTransport.cpp:383)
   by 0x51BB5B3: epics::pvAccess::BlockingUDPTransport::run() (blockingUDPTransport.cpp:240)
...
 Address 0x759e8c0 is 0 bytes inside a block of size 264 alloc'd
   at 0x4C2D63F: operator new(unsigned long) (vg_replace_malloc.c:334)
   by 0x51C4DF2: epics::pvAccess::SimpleChannelSearchManagerImpl::create(std::shared_ptr<epics::pvAccess::Context> const&) (simpleChannelSearchManagerImpl.cpp:46)
   by 0x5201BFB: (anonymous namespace)::InternalClientContextImpl::internalInitialize() (clientContextImpl.cpp:4161)
   by 0x5200F00: (anonymous namespace)::InternalClientContextImpl::initialize() (clientContextImpl.cpp:4064)
   by 0x5204277: epics::pvAccess::createClientProvider(std::shared_ptr<epics::pvAccess::Configuration> const&) (clientContextImpl.cpp:4829)
   by 0x51B7839: epics::pvAccess::(anonymous namespace)::FunctionFactory::sharedInstance() (ChannelAccessFactory.cpp:104)
   by 0x51B7205: epics::pvAccess::ChannelProviderRegistry::getProvider(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (ChannelAccessFactory.cpp:42)
   by 0x12481D: main (pvget.cpp:552)
 Block was alloc'd by thread #1
```

**Comment by mdavidsaver** [comment]
There is a chicken+egg game involving creating SimpleChannelSearchManagerImpl and the UDP search thread which is being resolved poorly (thread starts before allocation).

**Comment by mdavidsaver** [comment]
Should be fixed by aec1c7c9d7a0ab0f4ae913de2962c9d824ebc367 and 1a100a095513e76945426d5781a7506ef5bac28f

**Comment by mdavidsaver** [comment]
Released with Base 7.0.2
